File: ios-catalogue/App/Foundations/MotionShowcase.swift
import SwiftUI
import CoreUI

/// Tap a row to play that easing × duration token combination.
/// Demonstrates that motion tokens drive real Animation curves on iOS.
struct MotionShowcase: View {
    @Environment(\.colorScheme) private var scheme

    private static let rows: [(String, (Double, Double, Double, Double), Double)] = [
        ("standard / default", PrismaMotion.Easing.standard, PrismaMotion.Duration.default),
        ("decelerate / default", PrismaMotion.Easing.decelerate, PrismaMotion.Duration.default),
        ("accelerate / fast", PrismaMotion.Easing.accelerate, PrismaMotion.Duration.fast),
        ("emphasized / slow", PrismaMotion.Easing.emphasized, PrismaMotion.Duration.slow),
        ("spring / slower", PrismaMotion.Easing.spring, PrismaMotion.Duration.slower),
        ("linear / default", PrismaMotion.Easing.linear, PrismaMotion.Duration.default)
    ]

    var body: some View {
        VStack(alignment: .leading, spacing: PrismaSpacing.sp4) {
            Text("Tap a row to play the animation.")
                .font(PrismaTypography.bodySm.font)
                .foregroundStyle(PrismaSemanticColors.textTertiary.themed(scheme))

            ForEach(Self.rows.indices, id: \.self) { idx in
                let (name, easing, duration) = Self.rows[idx]
                MotionRow(name: name, easing: easing, durationSeconds: duration)
            }
        }
    }
}

private struct MotionRow: View {
    let name: String
    let easing: (Double, Double, Double, Double)
    let durationSeconds: Double

    @Environment(\.colorScheme) private var scheme
    @State private var play = false

    private var dotOffset: CGFloat { play ? 200 - 16 : 0 }

    var body: some View {
        Button(action: trigger) {
            HStack(alignment: .center, spacing: PrismaSpacing.sp4) {
                VStack(alignment: .leading, spacing: 2) {
                    Text(name)
                        .font(.system(size: 13, weight: .medium, design: .monospaced))
                        .foregroundStyle(PrismaSemanticColors.textPrimary.themed(scheme))
                    Text("\(Int(durationSeconds * 1000))ms")
                        .font(.system(size: 11, design: .monospaced))
                        .foregroundStyle(PrismaSemanticColors.textTertiary.themed(scheme))
                }
                .frame(width: 180, alignment: .leading)

                ZStack(alignment: .leading) {
                    Capsule()
                        .fill(PrismaSemanticColors.borderSubtle.themed(scheme))
                        .frame(width: 200, height: 4)

                    Circle()
                        .fill(PrismaSemanticColors.accentDefault.themed(scheme))
                        .frame(width: 16, height: 16)
                        .offset(x: dotOffset)
                }

                Spacer()
            }
            .padding(PrismaSpacing.sp4)
            .background(PrismaSemanticColors.surfaceSunken.themed(scheme))
            .clipShape(RoundedRectangle(cornerRadius: PrismaRadius.md))
        }
        .buttonStyle(.plain)
    }

    private func trigger() {
        let curve = Animation.timingCurve(easing.0, easing.1, easing.2, easing.3, duration: durationSeconds)
        withAnimation(curve) { play.toggle() }
        // Auto-reset after duration so the next tap replays cleanly.
        DispatchQueue.main.asyncAfter(deadline: .now() + durationSeconds + 0.1) {
            withAnimation(.linear(duration: 0)) { play = false }
        }
    }
}

File: android-catalogue/catalogue/src/main/java/xyz/ksharma/prisma/catalogue/foundations/MotionShowcase.kt
package xyz.ksharma.prisma.catalogue.foundations

import androidx.compose.animation.core.CubicBezierEasing
import androidx.compose.animation.core.animateFloatAsState
import androidx.compose.animation.core.tween
import androidx.compose.foundation.background
import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.graphics.graphicsLayer
import androidx.compose.ui.text.font.FontFamily
import androidx.compose.ui.unit.dp
import xyz.ksharma.prisma.coreui.themed
import xyz.ksharma.prisma.tokens.PrismaMotion
import xyz.ksharma.prisma.tokens.PrismaRadius
import xyz.ksharma.prisma.tokens.PrismaSemanticColors
import xyz.ksharma.prisma.tokens.PrismaSpacing
import xyz.ksharma.prisma.tokens.PrismaTypography

/** Tap a row to play that easing × duration token combination. */
@Composable
public fun MotionShowcase() {
    Column(
        modifier = Modifier.fillMaxWidth(),
        verticalArrangement = Arrangement.spacedBy(PrismaSpacing.Sp4),
    ) {
        Text(
            text = "Tap a row to play the animation.",
            style = PrismaTypography.BodySm,
            color = PrismaSemanticColors.TextTertiary.themed(),
        )

        MotionRow("standard / default", PrismaMotion.Easing.Standard,    PrismaMotion.Duration.Default)
        MotionRow("decelerate / default", PrismaMotion.Easing.Decelerate,  PrismaMotion.Duration.Default)
        MotionRow("accelerate / fast",    PrismaMotion.Easing.Accelerate,  PrismaMotion.Duration.Fast)
        MotionRow("emphasized / slow",    PrismaMotion.Easing.Emphasized,  PrismaMotion.Duration.Slow)
        MotionRow("spring / slower",      PrismaMotion.Easing.Spring,      PrismaMotion.Duration.Slower)
        MotionRow("linear / default",     PrismaMotion.Easing.Linear,      PrismaMotion.Duration.Default)
    }
}

@Composable
private fun MotionRow(
    name: String,
    easingPoints: FloatArray,
    durationMs: Int,
) {
    var play by remember { mutableStateOf(false) }
    val progress by animateFloatAsState(
        targetValue = if (play) 1f else 0f,
        animationSpec = tween(
            durationMillis = durationMs,
            easing = CubicBezierEasing(easingPoints[0], easingPoints[1], easingPoints[2], easingPoints[3]),
        ),
        finishedListener = { play = false },
        label = name,
    )

    Row(
        verticalAlignment = Alignment.CenterVertically,
        modifier = Modifier
            .fillMaxWidth()
            .clip(RoundedCornerShape(PrismaRadius.Md))
            .background(PrismaSemanticColors.SurfaceSunken.themed())
            .clickable { play = true }
            .padding(PrismaSpacing.Sp4),
    ) {
        Column(modifier = Modifier.padding(end = PrismaSpacing.Sp4)) {
            Text(
                text = name,
                style = PrismaTypography.LabelMd.copy(fontFamily = FontFamily.Monospace),
                color = PrismaSemanticColors.TextPrimary.themed(),
            )
            Text(
                text = "${durationMs}ms",
                style = PrismaTypography.LabelSm,
                color = PrismaSemanticColors.TextTertiary.themed(),
            )
        }

        // Track + animated dot
        Box(
            modifier = Modifier
                .padding(start = PrismaSpacing.Sp4)
                .clip(RoundedCornerShape(PrismaRadius.Full))
                .background(PrismaSemanticColors.BorderSubtle.themed())
                .size(width = 200.dp, height = 4.dp),
            contentAlignment = Alignment.CenterStart,
        ) {
            Box(
                modifier = Modifier
                    .size(16.dp)
                    .graphicsLayer {
                        translationX = progress * (200f - 16f) * density
                    }
                    .clip(RoundedCornerShape(PrismaRadius.Full))
                    .background(PrismaSemanticColors.AccentDefault.themed()),
            )
        }
    }
}
